tsconfig.json

```diff
@@ -11,7 +11,7 @@
 		"allowJs": true,
 		"baseUrl": ".",
 		"paths": {
-			"@clayui/*": ["packages/clay-*/src"],
+			"@clayui/*": ["node_modules/@clayui/*/src"],
```

Comment:
Wont this defeat the purpose of the monorepo? I think if we do this, we won't be importing the local packages while developing, it would look for it in node_modules instead.

Reply:
I was worried about that too but had no problems testing locally, as we are still using the yarn workspace it generates a symbolic link from the packages in the project's `node_modules` root so we will still keep looking at the same files as before but when generating the build he will not add the types together.

Reply:
oh interesting, I didnt think about how yarn would impact that.

So how come this fixes the issue for generating types? It seems like either solution should work.

Reply:
Yeah, this is a bit curious but when we point to files outside `node_modules` as in the case for `packages/...` TypeScript was considering that it should compile its types also because it is outside `node_modules`, we could ignore those types adding in the `exclude` setting but that would break everything, now when we point inside `node_modules` TypeScript interprets it as a dependency even though we added it in the `exclude` setting will work fine and will not generate the types because it understands that it is a dependency.

Reply:
ah i see. hmm, this seems like a strange issue and something thats been solved before. Have you seen any other libraries do the same thing? I mostly just want to avoid config changes that "happen to work" without having followed a documented API approach.

Reply:
Sure, we haven't set it up yet to point to `node_modules`. This was broken before when we set TypeScript to generate types.

I've tried checking on other projects but never able to find a project similar to our style, as dependencies between packages and build in each package, usually only have one build that runs at the root.

I think it makes sense to point to `node_modules` since we are treating packages as dependencies on each other is as close as we can get to a Clay package usage environment.

Reply:
yeah I guess that makes sense to me as well. Im cool with this change, but we should just be aware in case there might be any strange behaviors with inter-package dependencies.

Reply:
> I've tried checking on other projects but never able to find a project similar to our style, as dependencies between packages and build in each package, usually only have one build that runs at the root.

I think [my toy project](https://github.com/wincent/js) might have those attributes (eg. [a package](https://github.com/wincent/js/tree/master/packages/stable-stringify) with [a dependency](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/packages/stable-stringify/package.json#L33) on [another package](https://github.com/wincent/js/tree/master/packages/is-object) in the monorepo; and each package has [it's own build](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/packages/is-object/package.json#L19) where I build both [declarations](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/Makefile#L17) and [JS files](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/Makefile#L21)).

To see what it is actually doing when I run `yarn run build` in one of the packages, you can see the commands that are run here:

```
yarn run v1.17.3
$ workspace-scripts build stable-stringify
env BABEL_ENV=production babel --root-mode upward --source-root packages/stable-stringify/src --ignore '**/__tests__/**/*' --relative -d ../lib --extensions .js,.ts packages/stable-stringify/src
tsc --emitDeclarationOnly --declaration --project tsconfig-declaration.json
Successfully compiled 1 file with Babel.
env BABEL_ENV=es babel --keep-file-extension --root-mode upward --source-root packages/stable-stringify/src --ignore '**/__tests__/**/*' --relative -d ../lib --extensions .js,.ts packages/stable-stringify/src
Successfully compiled 1 file with Babel.
find packages/stable-stringify/lib -name '*.ts' -not -name '*.d.ts' -exec sh -c 'mv -f "$0" "${0%.ts}.mjs"' {} \;
find packages/stable-stringify/src -name '*.js.flow' -exec bash -c 'cp "$0" "${0/\/src\///lib//}"' {} \;
find packages -path 'packages/*/src/*' -and -name '*.d.ts' -and -not -path '*/node_modules/*' -exec bash -c 'mv "$0" "${0/\/src\///lib//}"' {} \;
✨  Done in 6.01s.
```

[This is the `tsconfig.json`](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/tsconfig.json) used for the source build, and [this one](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/tsconfig-declaration.json) extends it and is used to build the declarations. Notably, [the "paths" configuration](https://github.com/wincent/js/blob/4a035a69590643840644c69794a39322354091f5/tsconfig.json#L13) does not need "node_modules" in it:

```
    "paths": {
      "@wincent/*": ["*/src"]
    }
```

So I am not sure why this hack is required in Clay; it should be possible without it.

Reply:
Hmm this is interesting I tried changing the "path" to:
```
"@clayui/*": ["*/clay-*/src"]
"@clayui/*": ["clay-*/src"]
```

I made other settings in `tsconfig.json` also to approximate your but still still giving error, I don't know if it has to do with your `workspace-scripts` (probably)... but Storybook breaks and VSCode also.

Reply:
> I don't know if it has to do with your `workspace-scripts` (probably)

`workspace-scripts` is just a "thin wrapper" that runs the commands I pasted above; ie:

```
# actual build
env BABEL_ENV=production babel --root-mode upward --source-root packages/stable-stringify/src --ignore '**/__tests__/**/*' --relative -d ../lib --extensions .js,.ts packages/stable-stringify/src

# emit declaration
tsc --emitDeclarationOnly --declaration --project tsconfig-declaration.json
```

(and now that I look at it, I see it is probably making `.d.ts` files for all packages in the monorepo — whoops — but the actual build is only running against one package)

Reply:
I think our build process could eventually be improved by adopting RFC #1567, but I think we can go about it safely. @bryceosterhaus If you're unsure about this, do some manual testing to make sure we're going the right way for now.

Reply:
I'm fine with doing something for now to alleviate the issue at hand. But maybe we could also open an additional issue to make sure we further address this in the future.

Reply:
ok, I'll go ahead with it. Issue created #2249.

Reply:
Cross-referencing [this comment](https://github.com/liferay/clay/issues/2235#issuecomment-519809702) about unwanted "node_modules" prefix getting auto-inserted by VSCode, possibly related to this change.

Reply:
hey @wincent this is weird, we haven't released any of these changes yet and it wasn't supposed to happen since VSCode doesn't read the `tsconfig` file, it reads the `types` property from the package.json file from package to identify where the types are.